src/sqlfluff/dialects/dialect_spark3.py

```diff
@@ -2289,15 +2289,7 @@ class JoinClauseSegment(BaseSegment):
 
     type = "join_clause"
 
-    match_grammar = StartsWith(
-        OneOf(
-            Ref("JoinTypeKeywords"),
-            Ref("JoinKeywords"),
-            Ref("NaturalJoinKeywords"),
-        ),
-    )
-
-    parse_grammar = OneOf(
+    match_grammar = OneOf(
```

Comment:
This is the bug. If you have `StartsWith` you need a terminator. Maybe we should make that mandatory?

In places where `StartsWith` is the same as what `parse_grammar` does, there's no point to it. And worse - there are bugs like this!

Reply:
I don't think we need to make it mandatory - but this setup made a lot more sense in before we made the improvements to partial matching. Under the hood - the parser will now effectively be looking for things which _start with_ the things below anyway.